gnucash/gnome-utils/gnc-account-sel.c

```diff
@@ -136,151 +227,453 @@ gnc_account_sel_class_init (GNCAccountSelClass *klass)
+    {
+        if (g_list_find_custom (gas->acctCommodityFilters,
+                                GINT_TO_POINTER(xaccAccountGetCommodity (acc)),
+                                gnc_commodity_compare_void) == NULL)
+        {
+            included = FALSE;
+        }
+    }
+    return included;
+}
 
-    g_signal_handlers_unblock_by_func (gas->combo, combo_changed_cb , gas);
+static gboolean
+visible_func (GtkTreeModel *model, GtkTreeIter *iter, gpointer user_data)
+{
+    GNCAccountSel *gas = GNC_ACCOUNT_SEL(user_data);
+    Account *acc;
+    gboolean visible = TRUE;
 
-    g_list_free (atnd.outList);
-    if (currentSel)
-        g_free (currentSel);
+    gtk_tree_model_get (GTK_TREE_MODEL(gas->store), iter, ACCT_COL_PTR, &acc, -1);
+
+    if (acc)
+    {
+        visible = account_is_included (gas, acc);
+
+        if (gas->hide_placeholder && xaccAccountGetPlaceholder (acc))
+            visible = FALSE;
+
+        if (gas->hide_placeholder && xaccAccountIsHidden (acc))
+            visible = FALSE;
+    }
+    return visible;
 }
 
 static void
-gas_filter_accounts (gpointer data, gpointer user_data)
+row_deleted_in_store_cb (GtkTreeModel *model, GtkTreePath *path, gpointer user_data)
 {
-    account_filter_data *atnd;
-    Account *a;
+    GNCAccountSel *gas = GNC_ACCOUNT_SEL(user_data);
 
-    atnd = (account_filter_data*)user_data;
-    a = (Account*)data;
-    /* Filter as we've been configured to do. */
-    if (atnd->gas->acctTypeFilters)
+    if (gas->selected_account_ref)
     {
-        /* g_list_find is the poor-mans '(member ...)', especially
-         * easy when the data pointers in the list are just casted
-         * account type identifiers. */
-        if (g_list_find (atnd->gas->acctTypeFilters,
-                          GINT_TO_POINTER(xaccAccountGetType (a)))
-                == NULL)
+        GtkTreePath *selected_account_path = gtk_tree_row_reference_get_path (gas->selected_account_ref);
+
+        if (selected_account_path == NULL) // path is already invalid after row delete
         {
-            return;
+            GtkEntry *entry = GTK_ENTRY(gtk_bin_get_child (GTK_BIN(gas->combo)));
+
+            g_signal_handlers_block_by_func (gas->combo, combo_changed_cb , gas);
+            gtk_combo_box_set_active (GTK_COMBO_BOX(gas->combo), -1);
+            gtk_editable_delete_text (GTK_EDITABLE(entry), 0, -1);
+            gtk_tree_row_reference_free (gas->selected_account_ref);
+            gas->selected_account_ref = NULL;
+            g_signal_emit_by_name (gas, "account_sel_changed");
+            g_signal_handlers_unblock_by_func (gas->combo, combo_changed_cb , gas);
         }
+        gtk_tree_path_free (selected_account_path);
     }
+}
 
-    if (atnd->gas->acctCommodityFilters)
+static void
+row_changed_in_store_cb (GtkTreeModel *model, GtkTreePath *path, GtkTreeIter *iter, gpointer user_data)
+{
+    GNCAccountSel *gas = GNC_ACCOUNT_SEL(user_data);
+
+    if (gas->selected_account_ref)
     {
```

Comment:
General comment applies pretty much everywhere: Deeply indented code is hard to read so if the function is done when the condition is false then reverse the condition and return. e.g.
```
    if (!gas->selected_account_ref)
        return;
```

Reply:
OK.